[RFR] Adds Redfish API client

matejart opened the pull request:
With this PR we replace the hard-coded value for the stat check with that obtained by the Redfish provider's implementation now uses [the Redfish API client](https://pypi.org/project/redfish-client/).

We initiated a new entity `Server` and contributed `ServerStats` enum, which `RedfishServer` implements with its API specifics. The `RedfishSystem` is implemented such that it can be used both for testing stats at the provider level and for a selected physical server instance.

~~Depends on https://github.com/ManageIQ/integration_tests/pull/7735 and https://github.com/ManageIQ/wrapanapi/pull/331.~~

/cc @tadeboro, @miha-plesko

dajoRH commented:
Lint report for commit 05e606d080916bbc8ff64a0a474811026bf09645:

`wrapanapi/systems/redfish.py`:
- :red_circle: Line 8:1: **F401** *'json' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

matejart commented:
I now added the stat checks for the physical server.

tadeboro commented on wrapanapi/systems/redfish.py:
```diff
@@ -15,16 +17,30 @@ class RedfishSystem(System):
+        'power_state': lambda self, requester: self.get_server_power_status(requester.name),
+    }
+
+    def __init__(self, hostname, username, password, security_protocol, api_port=443, **kwargs):
```
`security_protocol` became required parameter and this means that doctring for this class should be updated accordingly.

tadeboro commented on wrapanapi/systems/redfish.py:
```diff
@@ -35,3 +51,84 @@ def info(self):
+
+    def server_stats(self, *requested_stats, **kwargs):
+        # Get the requester which represents the class of this method's caller
+        requester = kwargs.get('requester')
```
Nitpick: last three lines can be simplified to `def server_stats(self, requester=None, *requested_stats, **_)`. It would be even better to require `requester` as a first, non-optional parameter, but that would mean breaking the API and fixing all the sites that call this method.

tadeboro replied:
Did you try to run it and it failed? Because looking at the method call, things should work as I described. In any case, this was just a nitpick, so feel free to ignore it.

tadeboro replied:
Sorry, my bad. You are right, suggested refactoring will probably error-out because `requester` parameter will have multiple values assigned. There is no sane way of handling this under python 2.

matejart replied:
My approach is even more radical: changing the call directly:

https://github.com/ManageIQ/integration_tests/pull/7735
https://github.com/ManageIQ/wrapanapi/pull/331

tadeboro commented on wrapanapi/systems/redfish.py:
```diff
@@ -35,3 +51,84 @@ def info(self):
+
+    def server_inventory(self, *requested_items, **kwargs):
+        # Get the requester which represents the class of this method's caller
+        requester = kwargs.get('requester')
```
Same as `server_stats` definition.

tadeboro commented on wrapanapi/systems/redfish.py:
```diff
@@ -35,3 +51,84 @@ def info(self):
+        for proc in physical_system.Processors.Members:
+            num_cores += int(proc.TotalCores)
+
+        return num_cores
```
Nitpick: this can be simplified into something similar to `sum([int(p.TotalCores) for p in physical_system.Processors.Members])`.

matejart replied:
Thanks, I applied the pattern you suggested.

tadeboro commented on wrapanapi/systems/redfish.py:
```diff
@@ -35,3 +53,80 @@ def info(self):
+            # Redfish representation of a computation system doesn't have a
+            # name, and the provider uses the Id of the system as a name
+            if node.Id == str(server_name):
+                return node
```
Redfish API client already handles caching for us, so calling `self.api_client.Systems.Members` will only fetch data from remote on first use.

miha-plesko replied:
👍 fancy :)

@matejart in such case I suggest we override `disconnect` method on RedfishSystem and clear cache in it. Not necessarily in this PR though.

matejart replied:
When the garbage collector destroys the RedfishSystem instance, it will also destroy the client and the along with it the cache. Hence, the `__del__` method is redundant. I don't think any explicit disconnecting is needed.

miha-plesko replied:
Ye ye, I was thinking more like if you use it twice:

```python
mgmt = provider.get_mgmt()
mgmt.validate_state()

# create new server or something

mgmt.validate_state() #<--- this one will use cached list of servers so it will fail
```

Unless @tadeboro if the client is smart enough to automatically update the cache when new server is created?

matejart commented on wrapanapi/systems/redfish.py:
```diff
@@ -13,18 +15,34 @@ class RedfishSystem(System):
+    _server_stats_available = {
+        'cores_capacity': lambda self, requester: self.get_server_cores(requester.name),
+        'memory_capacity': lambda self, requester: self.get_server_memory(requester.name),
+        'num_firmwares': lambda self, requester: self.get_num_server_firmwares(requester.name),
```
I had to add them because [the test implementation expects them](https://github.com/ManageIQ/integration_tests/blob/58e8401f96ec3f5fe23ad1b6ac6b508c3e6f98ae/cfme/physical/physical_server.py#L62-L63).

matejart commented on wrapanapi/systems/redfish.py:
```diff
@@ -35,3 +53,80 @@ def info(self):
+
+        return physical_system.MemorySummary.TotalSystemMemoryGiB * 1024
+
+    def get_num_server_firmwares(self, server_name):
```
As explained, I need the stubs, so I might just as well name them as likely future methods.

matejart commented:
Thank you for the comments, I applied all the pertinent ones.

miha-plesko commented on wrapanapi/systems/redfish.py:
```diff
@@ -4,6 +4,8 @@
 """
 from __future__ import absolute_import
 
+import redfish_client
```
Sorry, only realized now: there should be no empty line between this one and `from __funture__ ...` above according to guidelines https://cfme-tests.readthedocs.io/en/master/guides/dev_guide.html#code-style

matejart replied:
I think that in this case the space should be OK, because `redfish_client` is in a different group than `from _future__`.

matejart replied:
I think the space here should be ok because `redfish_client` belongs in a different group of imports.

miha-plesko commented on wrapanapi/systems/redfish.py:
```diff
@@ -13,18 +15,54 @@ class RedfishSystem(System):
+            the API. Expected values: 'Non-SSL', 'SSL', 'SSL without validation'
     """
+    # Power states as defined in Redfish
+    RF_POWER_STATE_ON = 'On'
```
Just to be sure: we need those variables defined as standalones to make them accesible from integration_tests, right? If not, I'd just provide the hash directly like

```python
POWER_STATE_MAPPING = {
  'On': 'on',
  'Off': 'off',
  ...
}
```

because we're polluting the code :) Alternatively we could put them in separate file say `wrapanapi/utils/const_redfish.py`.

matejart replied:
Yes, I foresee tests where we will want to check the status in API, not just in GUI.

mshriver commented on wrapanapi/systems/redfish.py:
```diff
@@ -13,18 +15,54 @@ class RedfishSystem(System):
+    RF_POWER_STATE_OFF = 'Off'
+    RF_POWER_STATE_POWERING_ON = 'PoweringOn'
+    RF_POWER_STATE_POWERING_OFF = 'PoweringOff'
+    # Power states in CloudForms
```
These should be defined in the CloudForms classes for modeling RedFish, not here in wrapanapi.

If anything, the states that RedFish returns should be mapped to VmState enumerations, as demonstrated here in the EC2System class:
https://github.com/ManageIQ/wrapanapi/blob/master/wrapanapi/systems/ec2.py#L47

matejart replied:
> These should be defined in the CloudForms classes for modeling RedFish, not here in wrapanapi.

Just to be clear, you are OK with the `RF_POWER_STATE_*` definitions (I did them because I've seen [your comment](https://github.com/ManageIQ/wrapanapi/pull/294#discussion_r207254977) in what I thought was a similar situation), but I should move the mapping CF ones somewhere else?

> If anything, the states that RedFish returns should be mapped to VmState enumerations, as demonstrated here in the EC2System class

These states are not about VMs, they are about physical servers. I don't think mixing the two would be a good idea. But unlike the VM representations, there is no such common representation that would be defined or even prescribed for the physical servers. So I wrote the mapping assuming something that isn't there - the providers actually push the lower-case version of whatever state is there and that appears on the GUI.

Assuming that we decide to establish a common set of states, would it then make sense for us to create a `wrapanap/entities/physical_server.py` with a `PhysicalServerState` enumeration?

mshriver replied:
You're probably right that its a bad idea to conflate VM States and physical server states, and are right about creating a base entity class for servers. I would recommend just using `server.py` and ServerState class.

Then a RedfishServer class can inherit this entity class. 

Note how `state_map` is used in `entities.vm.Vm`, and the `_api_state_to_vmstate()` method resolves the map to a constant state.  The system-specific VM classes then implement `_get_state`, calling `_api_state_to_vmstate`  (note this is an abstractmethod for `Vm`).

The pattern described above should be mirrored in a new entity class, which may change how you've defined `FR_POWER_STATE_*` here.

So by extending the entity base class into a new server entity class, you can establish the same structure and pattern that is used by the VM's to manage server states in a consistent manner across different provider types that host servers.

I suggest putting this in a 'Server' class and not 'physical_server' because this same class could be applied to hosts for other providers (rhevm, openstack, virtualcenter, etc)

mshriver replied:
If the states are essentially the same, I think we'd be better off abstracting VmState a bit more to a generic 'PowerState' class, and then extending that as a 'mixin' type class where we need to add some state unique to a Vm, or a server, or a template, or a switch, etc.

mshriver commented on wrapanapi/systems/redfish.py:
```diff
@@ -33,5 +71,80 @@ def _identifying_attrs(self):
+        """
+        return self.get_server(resource_id).PowerState
+
+    def get_server_power_status_translated(self, resource_id):
```
This method shouldn't exist here in wrapanapi, this has to do with MIQ specifically. VmState enumerations are what should be used to standardize mgmt class states within MIQ's integration_tests.

matejart replied:
👍

mshriver commented on wrapanapi/systems/redfish.py:
```diff
@@ -33,5 +71,80 @@ def _identifying_attrs(self):
+        return 0
+
+    def get_num_network_devices(self, resource_id):
+        # not currently supported
```
I don't think this is a good practice, these should raise NotImplementedError. Returning an integer is misleading.

matejart replied:
In principle I agree. I actually wasn't going to even include this method until we can support it. But the way the test framework for physical servers currently works it effectively forces a specific system's implementation to have an answer to certain stats. If I omit the method or raise NotImplementedError, the test will always fail.

I know the right answer would be to improve the way the integration tests perform the tests for the provider and the physical servers. But until then are we OK with the integration tests that fail because they don't fully overlap with the base class assumptions?

mshriver replied:
Ah! awesome. I was literally just typing a response, that you're right about updating tests to handle the NotImplementedError and skip, or removing the method call from `_server_stats_available` mapping.

mshriver commented on wrapanapi/systems/redfish.py:
```diff
@@ -33,5 +71,80 @@ def _identifying_attrs(self):
-    def __del__(self):
-        """Disconnect from the API when the object is deleted"""
+    def server_stats(self, *requested_stats, **kwargs):
+        # Get the requester which represents the class of this method's caller
```
Docblocks explaining the intent of this and `server_inventory` methods (difference of counts, are they returning dicts, tuples, objects, names, etc.

mshriver commented on wrapanapi/systems/redfish.py:
```diff
@@ -33,5 +71,80 @@ def _identifying_attrs(self):
+
+    def get_server(self, resource_id):
+        """
+        Fetch a Redfish resource that represents a particular physical server.
```
:+1: good docblock

mshriver commented on wrapanapi/systems/redfish.py:
```diff
@@ -33,5 +71,80 @@ def _identifying_attrs(self):
+    def num_servers(self):
+        return len(self.api_client.Systems.Members)
+
+    def get_server_cores(self, resource_id):
```
All of these methods take a resource_id in order to identify the server - when there's already a get_server method.

I think that you should implement a RedfishServer class that extents wrapanapi.entity class (like VMs, templates for other providers)

This way `get_server` would return a `RedfishServer` object, that then contains the `get_cores`, `get_memory`, etc methods. They can all reference `self.raw`, which should be the RedFish api_client object with the raw parameters.

The pattern described above is one of the core design principles of wrapanapi 3.0.

mshriver replied:
Here is where the `entities.server.Server` class would come in handy, with a specific `RedfishServer` defined locally.

matejart commented:
@mshriver thank you for your comments. I have addressed most of the issues. I will create a `server.py` entity as you suggested. But maybe that can go into the next PR so that this will unblock the [PR](https://github.com/ManageIQ/integration_tests/pull/7726) that depends on this PR?

miha-plesko commented on wrapanapi/entities/server.py:
```diff
@@ -0,0 +1,107 @@
+            return ServerState.UNKNOWN
+
+    @abstractmethod
+    def _get_state(self):
```
Can I suggest to replace `_get_state` with two functions:

```python
@property
def current_state(self)
  return self._api_state_to_serverstate(self.current_api_state)

@abstractproperty
def current_api_state(self):
  # to be implemented in subclass - it returns current server state from the API
```

I prefer `current_state` name because it better reflects what the function does - it returns current state. Also, we don't want every subclass to be calling the `_api_state_to_serverstate` if we can do it here already.

matejart replied:
We can't do that, because `_get_state` is prescribed by the ancestor class, so we have to implement it.

matejart replied:
[deleted]
looks like that ancestor class is what I define. I'll try something different then.

miha-plesko commented on wrapanapi/systems/redfish.py:
```diff
@@ -4,27 +4,124 @@
 
+import redfish_client
+
+from wrapanapi.entities.server import Server, ServerState
```
Not sure why we're listing Server and ServerState in `wrapanapi/entities/__init__.py` if we then import them directly here? :) I suggest you either go with

```python
from wrapanapi.entities import Server, ServerState
```
or just remove the two classes from `wrapanapi/entities/__init__.py` (I prefer the second option TBH).

matejart replied:
👍 I went for the first option, since all the others are already in the `__init__.py`

miha-plesko commented on wrapanapi/systems/redfish.py:
```diff
@@ -4,27 +4,124 @@
+        return self.raw.MemorySummary.TotalSystemMemoryGiB * 1024
+
+    @property
+    def state(self):
```
According to my suggestion above, I'd rename this function into `def current_api_state(self)` and BOOOM everything will work out of the box.

matejart replied:
Since we are talking about possibly going to a common interface for both `Vm` and `Server` classes, I'll keep their schema for now.

miha-plesko commented on wrapanapi/systems/redfish.py:
```diff
@@ -4,27 +4,124 @@
+        """Retrieve the current power status of the physical server."""
+        return self.raw.PowerState
+
+    def _get_state(self):
```
We don't need this function in subclasses, please remove (see my suggestion above).

miha-plesko commented on wrapanapi/systems/redfish.py:
```diff
@@ -4,27 +4,124 @@
+        """
+        return self._api_state_to_serverstate(self.state)
+
+    def _identifying_attrs(self):
```
IMAO it would be better if we didn't copy-paste docstring from the interface, because we only pollute the code. I'd only add comments if we do something special.

miha-plesko replied:
Please remove redundant docstrings in all other interface implemetations as well.

matejart replied:
Actually those were for your benefit. :D
👍  I'll remove them.

miha-plesko replied:
PyCharm shows me a "go to ancestor definition" for each method so the documentation is always just a click away ;)

matejart replied:
Yes, but not everybody uses PyCharm. I put the comments back because they help with code understanding more than bother with too much text.

tadeboro commented on wrapanapi/entities/server.py:
```diff
@@ -0,0 +1,107 @@
+    def valid_states(cls):
+        return [
+            var_val for _, var_val in vars(cls).items()
+            if isinstance(var_val, basestring) and var_val.startswith('ServerState.')
```
Using `str` on python 2.7 in this check will miss all unicode strings, so I would advise against it. If we would like to be compatible with python 2 and 3, a bit more work is needed.

mshriver replied:
Please use `six` and `six.string_types` for py2+py3 compatibility.

matejart replied:
Done.

matejart commented on wrapanapi/entities/server.py:
```diff
@@ -0,0 +1,107 @@
+    # Implementations must define a dict which maps API states returned by the
+    # system to a ServerState. Example:
+    #    {'On': ServerState.ON, 'Off': ServerState.OFF}
+    state_map = None
```
This is an "abstract" field, so I guess the implementor of the class extending `Server` will have to supply its own constant to override it, while the base implementation still needs to check if it's a dict.

matejart commented on wrapanapi/entities/server.py:
```diff
@@ -0,0 +1,107 @@
+
+        Since abc has no 'abstract class property' concept, this is the approach taken.
+        """
+        state_map = getattr(self, 'state_map')
```
👍 
We don't need it.

matejart commented on wrapanapi/entities/server.py:
```diff
@@ -0,0 +1,107 @@
+        """
+        state_map = getattr(self, 'state_map')
+        if (not state_map or not isinstance(state_map, dict) or
+                not all(value in ServerState.valid_states() for value in state_map.values())):
```
I agree, this works too with less characters. But the original one has additional points in clarity what it does.

matejart commented on wrapanapi/entities/server.py:
```diff
@@ -0,0 +1,107 @@
+                "Unmapped Server state '%s' received from system, mapped to '%s'",
+                api_state, ServerState.UNKNOWN
+            )
+            return ServerState.UNKNOWN
```
We _want_ that `WARN` in the log if the dev has made a mistake with a mapping. So we don't really save anything by refactoring.

mshriver commented on wrapanapi/systems/redfish.py:
```diff
@@ -33,5 +130,61 @@ def _identifying_attrs(self):
+            requested_items: a list of named items to be checked
+        """
+        # Get the requester which represents the class of this method's caller
+        requester = kwargs.get('requester')
```
This appears to be required, as you later reference an attribute on `requester`. 

I don't see the need to take `kwargs` here, what's the reasoning behind this kwarg lookup instead of just a required `requester` parameter listed before `*requested_items`?

matejart replied:
Doing that doesn't work because of the way this method is called.
But please see my PRs https://github.com/ManageIQ/wrapanapi/pull/331 and https://github.com/ManageIQ/integration_tests/pull/7735 where I address this issue.

matejart replied:
This PR is for new Redfish-specific features. PR #331 and ManageIQ/integration_tests#7735 are changing a signature that results changes both in Lenovo's code and the Redfish provider. And the genesis of the PRs was really 1) add features to Redfish provider by using the Lenovo's code as example, and 2) address comments that required a deeper refactor later.

If need be, I can reorder the PRs, or I can put those commits here, but that will really muddy the purpose.

matejart replied:
Apologies, I didn't answer the last question: while this is a new method, it is called from [cfme/physical/physical_server.py](
https://github.com/ManageIQ/integration_tests/blob/2c39a4e47ca25c7e9e97e6ee3ab9d796b76a3b24/cfme/physical/physical_server.py#L240-L241), i.e. a piece of code in the integration tests that is common to all physical servers - Lenovo's xClarity, Redfish.

matejart commented:
I now flipped the dependencies so that the first PRs to look at are the ones that refactor integration tests for physical servers (and Lenovo): https://github.com/ManageIQ/wrapanapi/pull/331 and https://github.com/ManageIQ/integration_tests/pull/7735.

matejart commented:
I believe that I have addressed all the comments.

mshriver commented:
I believe your Travis failures will be resolved by #336

matejart commented:
> 
> 
> I believe your Travis failures will be resolved by #336

Thanks! I was looking at the errors and they seem unrelated to my changes.

mshriver commented:
@matejart I rebased after merging in order to trigger travis. Sorry to 'step on toes', just trying to push for release in time for integration_tests merge-freeze.
